sycl/include/CL/sycl/group.hpp

```diff
@@ -120,25 +121,134 @@ template <int Dimensions = 1> class group {
+  get_local_linear_id() const {
+    id<Dimensions> localId = get_local_id();
+    return localId[0] * groupRange[1] + localId[1];
+  }
+
+  template <int dims = Dimensions>
+  typename detail::enable_if_t<(dims == 3), size_t>
+  get_local_linear_id() const {
+    id<Dimensions> localId = get_local_id();
+    return (localId[0] * groupRange[1] * groupRange[2]) +
+           (localId[1] * groupRange[2]) + localId[2];
+  }
+
   range<Dimensions> get_local_range() const { return localRange; }
 
   size_t get_local_range(int dimension) const { return localRange[dimension]; }
 
+  template <int dims = Dimensions>
+  typename detail::enable_if_t<(dims == 1), size_t>
+  get_local_linear_range() const {
+    auto localRange = get_local_range();
+    return localRange[0];
+  }
+
+  template <int dims = Dimensions>
+  typename detail::enable_if_t<(dims == 2), size_t>
+  get_local_linear_range() const {
+    auto localRange = get_local_range();
+    return localRange[0] * localRange[1];
+  }
+
+  template <int dims = Dimensions>
+  typename detail::enable_if_t<(dims == 3), size_t>
+  get_local_linear_range() const {
+    auto localRange = get_local_range();
+    return localRange[0] * localRange[1] * localRange[2];
+  }
+
   range<Dimensions> get_group_range() const { return groupRange; }
 
   size_t get_group_range(int dimension) const {
     return get_group_range()[dimension];
   }
 
+  template <int dims = Dimensions>
+  typename detail::enable_if_t<(dims == 1), size_t>
+  get_group_linear_range() const {
+    auto groupRange = get_group_range();
+    return groupRange[0];
+  }
+
+  template <int dims = Dimensions>
+  typename detail::enable_if_t<(dims == 2), size_t>
+  get_group_linear_range() const {
+    auto groupRange = get_group_range();
+    return groupRange[0] * groupRange[1];
+  }
+
+  template <int dims = Dimensions>
+  typename detail::enable_if_t<(dims == 3), size_t>
+  get_group_linear_range() const {
+    auto groupRange = get_group_range();
+    return groupRange[0] * groupRange[1] * groupRange[2];
+  }
+
+  template <int dims = Dimensions>
+  typename detail::enable_if_t<(dims == 1), range<Dimensions>>
+  get_max_local_range() const {
+    return range<Dimensions>{
+        static_cast<size_t>(info::device::max_work_group_size)};
+  }
+
+  template <int dims = Dimensions>
+  typename detail::enable_if_t<(dims == 2), range<Dimensions>>
+  get_max_local_range() const {
+    return range<Dimensions>{
+        static_cast<size_t>(info::device::max_work_group_size),
+        static_cast<size_t>(info::device::max_work_group_size)};
+  }
+
+  template <int dims = Dimensions>
+  typename detail::enable_if_t<(dims == 3), range<Dimensions>>
+  get_max_local_range() const {
+    return range<Dimensions>{
+        static_cast<size_t>(info::device::max_work_group_size),
+        static_cast<size_t>(info::device::max_work_group_size),
+        static_cast<size_t>(info::device::max_work_group_size)};
+  }
```

Comment:
I appreciate the specification could be clearer here, but the intent of this function is to return the maximum number of work-items in any work-group for the kernel that is currently executing, rather than the maximum number of work-items  supported by the device.

Because we don't support non-uniform work-group sizes, I think you can replace all of this with:
```c++
size_t get_max_local_range() const {
  return get_local_range();
}
```

The reason both functions exist is that they return different values for the `sub_group` class.  Supporting both in `group` allows developers to write generic code accepting either a `sub_group` or ` group`.

Reply:
Done